### ISSUE TITLE
[Bug] Date objects not supported in query parameters

### ISSUE THREAD

**Issue by yhnavein** [suhaotian/xior#23]
Let's consider the following code:

```ts
import xior, { type XiorResponse, type XiorRequestConfig } from 'xior';

const filter = {
  ids: [1, 2, 3],
  dates: [new Date(), new Date()],
  dateFrom: new Date(),
  dateTo: new Date(),
};

http.request<FilterTestResponse>({
    url: url,
    method: 'GET',
    params: {
      filter: filter,
    },
    ...$config,
});
```

If I execute it in a browser then only `ids` property will be serialized and present in the query, but `dates`, `dateFrom` and `dateTo` won't be there because dates are objects and they will be handled by `encodeParams` through a ` if (typeof value === 'object')` branch which is not expected. 

The fix I produced for my ASPNET custom serialized looks like this and it's working fine with both cases:

```ts
function encodeParams<T = any>(
  params: T,
  encodeURI = true,
  parentKey: string | null = null
): string {
  if (params === undefined || params === null) return '';
  const encodedParams: string[] = [];
  const encodeURIFunc = encodeURI ? encodeURIComponent : (v: string) => v;
  const encodeValue = (value: any) =>
    encodeURIFunc(value instanceof Date && !Number.isNaN(value) ? value.toISOString() : value);

  for (const key in params) {
    if (Object.prototype.hasOwnProperty.call(params, key)) {
      const value = (params as any)[key];
      if (value !== undefined) {
        const encodedKey = parentKey ? `${parentKey}.${encodeURIFunc(key)}` : encodeURIFunc(key);

        if (Array.isArray(value)) {
          // If the value is an array, encode each element individually
          value.forEach((element) => {
            encodedParams.push(`${encodeURIFunc(encodedKey)}=${encodeValue(element)}`);
          });
        } else if (value instanceof Date && !Number.isNaN(value)) {
          // If the value is a Date, convert it to ISO format
          encodedParams.push(`${encodeURIFunc(encodedKey)}=${encodeValue(value)}`);
        } else if (typeof value === 'object') {
          // If the value is an object or array, recursively encode its contents
          const result = encodeParams(value, encodeURI, encodedKey);
          if (result !== '') encodedParams.push(result);
        } else {
          // Otherwise, encode the key-value pair
          encodedParams.push(`${encodeURIFunc(encodedKey)}=${encodeValue(value)}`);
        }
      }
    }
  }

  return encodedParams.join('&');
}

export const http = xior.create({
  baseURL: '',
  paramsSerializer: encodeParams,
});
```

But please note to not use this code, because it already has support for different approach for query params serialization (i.e. `filter.ids=1&filter.ids=2&...` as opposed to your previous mechanism `filter[ids][0]=1&filter[ids][1]=2&...`

**Comment by suhaotian** [comment]
I wonder why `value instanceof Date && !Number.isNaN(value)`  here need `!Number.isNaN(value)` ? 

Looks like useless?

**Comment by suhaotian** [comment]
Fixed in v0.5.2!

Now if the value is `Date`, will call `value.toISOString()`. 

For `arrayFormat` ecnoding, you can custom `paramsSerializer`, for example:

```ts
import { Xior, encodeParams } from 'xior';

export const http = xior.create({
  baseURL: '',
  paramsSerializer: params => encodeParams(params, true, null, {allowDots: true, arrayFormat: 'repeat', serializeDate: d => d.toISOString()}),
});
```

For more changelog: https://github.com/suhaotian/xior/releases/tag/v0.5.2

Let me know if still have any issues.

**Comment by yhnavein** [comment]
> ```ts
> value instanceof Date && !Number.isNaN(value) ? value.toISOString() : value
> ```

I used wrong code by mistake. The actual check should be `value instanceof Date && !isNaN(value) ? value.toISOString() : value` but it was autocompleted by my linter ;/

`Number.isNaN` works different than `isNaN`. And `isNaN` is discouraged from use as it's unsafe.

I have added it there as a counter-measure for the following case:

```
const value = new Date('illegal');

value instanceof Date ? value.toISOString() : value; // -> Exception is thrown. Not good

// without it
value instanceof Date && !isNaN(value) ? value.toISOString() : value; // -> `Invalid Date`
```

But I think the best approach is to use this instead:
`value instanceof Date && value.getTime() ? value.toISOString() : value`

**Comment by suhaotian** [comment]
Nice!

Fixed in v0.5.3!

**Comment by yhnavein** [comment]
Wow! You are a champion! Thanks ❤️
